File: financial-statement-studio/apps/api/app/schemas/financial_statement.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class FinancialStatementLine(BaseModel):
    """
    One ledger account displayed in a financial statement section.
    """

    ledger_account_id: str | None
    account_code: str | None
    account_name: str
    report_category: str

    amount: Decimal

    is_calculated: bool = False


class FinancialStatementSection(BaseModel):
    """
    A grouped financial-statement section.

    Examples:
    - Current assets
    - Revenue
    - Administrative expenses
    """

    key: str
    title: str

    items: list[FinancialStatementLine]

    total: Decimal


class ProfitOrLossResponse(BaseModel):
    """
    Calculated Statement of Profit or Loss.
    """

    financial_report_id: str
    company_id: str
    currency: str

    period_start: date
    period_end: date

    sections: list[FinancialStatementSection]

    revenue: Decimal
    direct_costs: Decimal
    gross_profit: Decimal

    other_income: Decimal

    administrative_expenses: Decimal
    selling_distribution_expenses: Decimal

    operating_profit: Decimal

    finance_costs: Decimal
    profit_before_tax: Decimal

    taxation: Decimal
    profit_after_tax: Decimal

    generated_at: datetime


class StatementOfFinancialPositionResponse(BaseModel):
    """
    Calculated Statement of Financial Position.
    """

    financial_report_id: str
    company_id: str
    currency: str

    as_of: date

    sections: list[FinancialStatementSection]

    current_assets: Decimal
    non_current_assets: Decimal
    total_assets: Decimal

    current_liabilities: Decimal
    non_current_liabilities: Decimal
    total_liabilities: Decimal

    recorded_equity: Decimal
    current_year_profit: Decimal
    total_equity: Decimal

    total_liabilities_and_equity: Decimal

    accounting_equation_difference: Decimal
    is_balanced: bool

    generated_at: datetime

File: financial-statement-studio/apps/web/src/types/financial-statement.ts
export type MoneyValue =
  | string
  | number;

export type FinancialStatementLine = {
  ledger_account_id: string | null;
  account_code: string | null;
  account_name: string;
  report_category: string;
  amount: MoneyValue;
  is_calculated: boolean;
};

export type FinancialStatementSection = {
  key: string;
  title: string;
  items: FinancialStatementLine[];
  total: MoneyValue;
};

export type ProfitOrLossStatement = {
  financial_report_id: string;
  company_id: string;
  currency: string;

  period_start: string;
  period_end: string;

  sections: FinancialStatementSection[];

  revenue: MoneyValue;
  direct_costs: MoneyValue;
  gross_profit: MoneyValue;

  other_income: MoneyValue;

  administrative_expenses: MoneyValue;
  selling_distribution_expenses: MoneyValue;

  operating_profit: MoneyValue;

  finance_costs: MoneyValue;
  profit_before_tax: MoneyValue;

  taxation: MoneyValue;
  profit_after_tax: MoneyValue;

  generated_at: string;
};

export type StatementOfFinancialPosition = {
  financial_report_id: string;
  company_id: string;
  currency: string;

  as_of: string;

  sections: FinancialStatementSection[];

  current_assets: MoneyValue;
  non_current_assets: MoneyValue;
  total_assets: MoneyValue;

  current_liabilities: MoneyValue;
  non_current_liabilities: MoneyValue;
  total_liabilities: MoneyValue;

  recorded_equity: MoneyValue;
  current_year_profit: MoneyValue;
  total_equity: MoneyValue;

  total_liabilities_and_equity: MoneyValue;

  accounting_equation_difference: MoneyValue;
  is_balanced: boolean;

  generated_at: string;
};